# POGENOM #
POGENOM (POpulation GENOmics from Metagenomes) is a computer program that calculates several population genomic parameters for a genome in relation to a set of metagnome samples.

## Description ##
POGENOM takes as input a file of the variant call format (VCF). This is generated by mapping one or several metagenome samples against a reference genome with a read aligner and calling variants using a variant caller. POGENOM calculates the nucleotide diversity (ùõë) within each sample. If a multiple-sample vcf file is provided as input, the fixation index (FST) will also be calculated between each pair of samples. If, in addition to the VCF file, an annotation file of the General Feature Format (GFF) is provided as input, gene-wise ùõë and FST will also be calculated. If, further, a genetic code file is provided, gene-wise ùõë and FST will be calculated also at the amino acid level. Now also non-synonymous to synonymous polymorphism rates (pN/pS) will be calculated for each gene in each sample. POGENOM also calculates some of the above parameters for all samples collectively, treating them as a metasample. For gene- and loci-wise calculations, a genome sequence needs to be provided. This can either be included in the end of the GFF file (in fasta version) or be provided as a separate fasta file. 

## Documentation ##
Documentation on how to install and use POGENOM as well as the Input_POGENOM pipeline for generating input to POGENOM is hosted on [readthedocs (https://pogenom.readthedocs.io/en/latest/). 

## Citing POGENOM ##
POGENOM doesn't have a paper yet, meanwhile please cite it like this:
Andersson AF, Sj√∂qvist C (2017). POGENOM: population genomics from metagenomes. https://github.com/EnvGen/POGENOM.

