.. POGENOM


.. figure:: pogenomfig.png
   :width: 400px
   :align: center


POGENOM's documentation
=======================
POGENOM is a computer program that calculates several population genomic parameters for a genome in relation to a set of metagnome samples.

Description
===========
POGENOM takes as minimal input a file of the variant call format (VCF). This is generated by mapping one or several metagenome samples against a reference genome with a read aligner such as `Bowtie2 <http://bowtie-bio.sourceforge.net/bowtie2/>`__ or `MOSAIK <https://github.com/wanpinglee/MOSAIK>`__ and calling variants using a variant caller such as `GATK <https://github.com/broadinstitute/gatk>`__ or `Freebayes <https://github.com/ekg/freebayes>`__. You can use the :doc:`Input_POGENOM` to generate the input. POGENOM calculates the nucleotide diversity, ùúã, within each sample. If multiple samples have been mapped, the fixation index (FST) is calculated for all pairs of samples. If, in addition to the VCF file, an annotation file of the General Feature Format (GFF) is provided, gene-wise ùúã and FST will be calculated. If, further, the genome sequence is provided in the GFF file or in a separate FASTA file, amino acid frequencies will be calculated for each codon position in each gene in each sample, and gene-wise ùúã and FST will be calculated also at the amino acid level. Now also non-synonymous to synonymous polymorphism rates (pN/pS) will be calculated for each gene in each sample. A description of how the different population genomic parameters are calculated can be found `here <https://github.com/EnvGen/POGENOM/blob/master/POGENOM_parameter_calculations.pdf>`__.

Installation
============

Download the latest POGENOM distribution from https://github.com/EnvGen/POGENOM/releases and extract the files.
You need to have Perl installed on your computer to run POGENOM. When running POGENOM, either move to the directory where you have put the files or give the path to the files when running, i.e. `perl path/to/pogenom.pl ...`


Contents
========

.. toctree::
   :maxdepth: 4

   self
   POGENOM_Usage
   Input_POGENOM
   LICENSE

Citing POGENOM
==============

POGENOM doesn't have a paper yet, meanwhile please cite it like this:

Sj√∂qvist C, Delgado Zambrano LF, Alneberg J, Andersson AF (2020). POGENOM: population genomics from metagenomes.https://github.com/EnvGen/POGENOM.

If you use the Input_POGENOM pipeline, make sure to cite the included software/packages that are listed in its documentation.
