Usage
=====

POGENOM takes as minimal input a file of the variant call format (VCF). This is generated by mapping one or several metagenome samples against a reference genome with a read aligner such as `Bowtie2 <http://bowtie-bio.sourceforge.net/bowtie2/>`__ or `MOSAIK <https://github.com/wanpinglee/MOSAIK>`__ and calling variants using a variant caller such as `GATK <https://github.com/broadinstitute/gatk>`__ or `Freebayes <https://github.com/ekg/freebayes>`__. You can use the :doc:`Input_POGENOM` to generate the input. POGENOM calculates the nucleotide diversity, ùúã, within each sample. If multiple samples have been mapped, the fixation index (FST) is calculated for all pairs of samples. If, in addition to the VCF file, an annotation file of the General Feature Format (GFF) is provided, gene-wise ùúã and FST will be calculated. If, further, the genome sequence is provided in the GFF file or in a separate FASTA file, amino acid frequencies will be calculated for each codon position in each gene in each sample, and gene-wise ùúã and FST will be calculated also at the amino acid level. Now also non-synonymous to synonymous polymorphism rates (pN/pS) will be calculated for each gene in each sample.
 
Minimum input::

    perl pogenom.pl --vcf_file VCF_FILE --out OUTPUT_FILES_PREFIX --genome_size GENOME_SIZE

or::

    perl pogenom.pl --vcf_file VCF_FILE --out OUTPUT_FILES_PREFIX --gff_file GFF_FILE

or::

    perl pogenom.pl --vcf_file VCF_FILE --out OUTPUT_FILES_PREFIX --fasta_file FASTA_FILE

POGENOM will calculate genome-wide ùúã and FST, if ``VCF_FILE`` is a multi-sample VCF.


If a GFF file is provided, gene-wise ùúã and gene-wise FST will also be calculated::

    perl pogenom.pl --vcf_file VCF_FILE --out OUTPUT_FILES_PREFIX --gff_file GFF_FILE
    
And if a genetic code file is provided, amino acid frequencies will be calculated for each codon position in each gene in each sample, and gene-wise ùúã and FST will be calculated also at the amino acid level. Now also non-synonymous to synonymous polymorphism rates (pN/pS) will be calculated for each gene in each sample.::

    perl pogenom.pl --vcf_file VCF_FILE --out OUTPUT_FILES_PREFIX --gff_file GFF_FILE --genetic_code_file GENETIC_CODE_FILE
    

Required arguments
^^^^^^^^^^^^^^^^^^


--vcf_file VCF_FILE                   
 Specify VCF file with data from a single or multiple samples.

--out OUTPUT_FILES_PREFIX             
 Specify the prefix of the output file name(s) (overwrites existing files with same names).

--genome_size GENOME_SIZE             
 Specify genome size (in bp; integer). Not required if `--gff_file` or `--fasta_file` with genome sequence is given.


Optional arguments
^^^^^^^^^^^^^^^^^^


--gff_file GFF_FILE                   
 Specify gff file. Either this, `--genome_size` or `--fasta_file` must be given.

--fasta_file FASTA_FILE
 Specify fasta file. Either this, `--genome_size` or `--gff_file` must be given

--genetic_code_file GENETIC_CODE_FILE
 Specify genetic code file. E.g. `standard_genetic_code.txt` in the POGENOM distribution.

--loci_file LOCI_FILE
 Specify file with ids of loci to include.

--sample_file SAMPLE_FILE   
Specify file with ids of samples to be included.

--min_count MIN_COUNT
 Specify minimum coverage for a locus to be included for the sample.

--min_found MIN_FOUND_IN
 Specify minimum number samples that a locus needs to be present in to be included.

--subsample SUBSAMPLE
 Specify coverage level at which to subsample.

--keep_haplotypes
 If this is used, POGENOM will not split haplotypes into single-nucleotide variants, which is otherwise the default behaviour.

--vcf_format
Specify VCF file format version. Can be set to freebayes (default) or GATK.
 
--fst_perm FST_PERM         
Specify number of permutations (integer) for making randomised gene-wise Fst. Without setting this randomised Fst are not generated. Warning: use with care, output files can become huge

--pi_only                   
Set this to make POGENOM only calculate and output genome-wide pi (fast)

--help
 To print help message on screen.

